File: imk-helper/Sources/ImkHelperCore/Protocol.swift
// Wire protocol mirror of src-tauri/src/input/imk/protocol.rs.
// MUST stay in lockstep with the Rust side — the snake_case JSON tags are
// the contract. Add a roundtrip test on both sides whenever you change a
// variant or field.

import Foundation

public struct ImkRange: Codable, Equatable {
    public let location: UInt32
    public let length: UInt32
    public init(location: UInt32, length: UInt32) {
        self.location = location
        self.length = length
    }
}

public struct ImkRect: Codable, Equatable {
    public let x: Double
    public let y: Double
    public let w: Double
    public let h: Double
    public init(x: Double, y: Double, w: Double, h: Double) {
        self.x = x; self.y = y; self.w = w; self.h = h
    }
}

/// Requests the Tauri side sends to the helper. Decoded server-side.
public enum ImkRequest: Equatable {
    case insert(text: String, replacementRange: ImkRange?)
    case getContext
    case ping
}

extension ImkRequest: Codable {
    // Manual Codable because `enum with associated values` + `serde tag =
    // "type"` don't line up via the synthesized derivation.
    private enum CodingKeys: String, CodingKey {
        case type, text, replacement_range
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let type = try c.decode(String.self, forKey: .type)
        switch type {
        case "insert":
            let text = try c.decode(String.self, forKey: .text)
            let range = try c.decodeIfPresent(ImkRange.self, forKey: .replacement_range)
            self = .insert(text: text, replacementRange: range)
        case "get_context":
            self = .getContext
        case "ping":
            self = .ping
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: c,
                debugDescription: "unknown request type '\(type)'"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .insert(text, range):
            try c.encode("insert", forKey: .type)
            try c.encode(text, forKey: .text)
            if let range { try c.encode(range, forKey: .replacement_range) }
        case .getContext:
            try c.encode("get_context", forKey: .type)
        case .ping:
            try c.encode("ping", forKey: .type)
        }
    }
}

/// Responses the helper sends back. Encoded server-side.
public enum ImkResponse: Equatable {
    case ok
    case context(
        beforeCursor: String,
        selectedText: String,
        cursorRect: ImkRect,
        appBundleId: String,
        appName: String
    )
    case error(message: String)
}

extension ImkResponse: Codable {
    private enum CodingKeys: String, CodingKey {
        case type, message
        case before_cursor, selected_text, cursor_rect, app_bundle_id, app_name
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let type = try c.decode(String.self, forKey: .type)
        switch type {
        case "ok":
            self = .ok
        case "context":
            self = .context(
                beforeCursor: try c.decode(String.self, forKey: .before_cursor),
                selectedText: try c.decode(String.self, forKey: .selected_text),
                cursorRect: try c.decode(ImkRect.self, forKey: .cursor_rect),
                appBundleId: try c.decode(String.self, forKey: .app_bundle_id),
                appName: try c.decode(String.self, forKey: .app_name)
            )
        case "error":
            self = .error(message: try c.decode(String.self, forKey: .message))
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: c,
                debugDescription: "unknown response type '\(type)'"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .ok:
            try c.encode("ok", forKey: .type)
        case let .context(beforeCursor, selectedText, cursorRect, appBundleId, appName):
            try c.encode("context", forKey: .type)
            try c.encode(beforeCursor, forKey: .before_cursor)
            try c.encode(selectedText, forKey: .selected_text)
            try c.encode(cursorRect, forKey: .cursor_rect)
            try c.encode(appBundleId, forKey: .app_bundle_id)
            try c.encode(appName, forKey: .app_name)
        case let .error(message):
            try c.encode("error", forKey: .type)
            try c.encode(message, forKey: .message)
        }
    }
}

File: src-tauri/src/input/imk/protocol.rs
//! Wire protocol for talking to the IMK helper.
//!
//! Newline-delimited JSON over Unix socket. Both sides MUST keep the
//! Swift `Protocol.swift` in `imk-helper/` in lockstep with this file —
//! the snake_case tags are normative.

use serde::{Deserialize, Serialize};

/// NSRange-equivalent: location + length, both in UTF-16 code units (what
/// macOS text APIs work in). 0-based location.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Range {
    pub location: u32,
    pub length: u32,
}

/// Screen-coordinate rectangle for the cursor. Origin is bottom-left
/// (Cocoa convention) — front-end converts to top-left as needed.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct Rect {
    pub x: f64,
    pub y: f64,
    pub w: f64,
    pub h: f64,
}

/// Messages the Tauri side sends to the IMK helper.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Request {
    /// Insert `text` at the current insertion point. If
    /// `replacement_range` is `Some`, replace that range first (used for
    /// the selected-text-transform workflow).
    Insert {
        text: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        replacement_range: Option<Range>,
    },

    /// Read the current cursor context: text before cursor, selected
    /// text, cursor rect, active app. Used to feed context into the
    /// polish prompt.
    GetContext,

    /// Liveness probe — server replies `Ok`. Used by reconnect logic.
    Ping,
}

/// Messages the IMK helper sends back to the Tauri side.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Response {
    /// Operation completed (no payload).
    Ok,

    /// Reply to `GetContext`.
    Context {
        before_cursor: String,
        selected_text: String,
        cursor_rect: Rect,
        app_bundle_id: String,
        app_name: String,
    },

    /// Operation failed. `message` is human-readable, not for parsing.
    Error { message: String },
}

#[cfg(test)]
mod tests {
    use super::*;

    fn roundtrip<T: Serialize + serde::de::DeserializeOwned + PartialEq + std::fmt::Debug>(v: T) {
        let s = serde_json::to_string(&v).expect("serialize");
        let back: T = serde_json::from_str(&s).expect("deserialize");
        assert_eq!(v, back, "roundtrip mismatch via {s}");
    }

    #[test]
    fn insert_without_range_omits_field() {
        let req = Request::Insert {
            text: "hello".into(),
            replacement_range: None,
        };
        let s = serde_json::to_string(&req).unwrap();
        // Critical: when no range, the field must be ABSENT (not null) so
        // Swift's Codable doesn't choke and the wire is compact.
        assert!(!s.contains("replacement_range"), "got: {s}");
        assert!(s.contains(r#""type":"insert""#));
        assert!(s.contains(r#""text":"hello""#));
    }

    #[test]
    fn insert_with_range_includes_field() {
        let req = Request::Insert {
            text: "rewrite".into(),
            replacement_range: Some(Range { location: 10, length: 5 }),
        };
        let s = serde_json::to_string(&req).unwrap();
        assert!(s.contains(r#""replacement_range":{"location":10,"length":5}"#), "got: {s}");
    }

    #[test]
    fn get_context_serializes_with_just_type() {
        let s = serde_json::to_string(&Request::GetContext).unwrap();
        assert_eq!(s, r#"{"type":"get_context"}"#);
    }

    #[test]
    fn ping_serializes() {
        let s = serde_json::to_string(&Request::Ping).unwrap();
        assert_eq!(s, r#"{"type":"ping"}"#);
    }

    #[test]
    fn ok_response_roundtrips() {
        roundtrip(Response::Ok);
    }

    #[test]
    fn error_response_carries_message() {
        let r = Response::Error { message: "no active client".into() };
        let s = serde_json::to_string(&r).unwrap();
        assert_eq!(s, r#"{"type":"error","message":"no active client"}"#);
        roundtrip(r);
    }

    #[test]
    fn context_response_roundtrips() {
        let r = Response::Context {
            before_cursor: "Hi John, regarding our meeting...".into(),
            selected_text: "".into(),
            cursor_rect: Rect { x: 320.0, y: 480.0, w: 2.0, h: 18.0 },
            app_bundle_id: "com.apple.mail".into(),
            app_name: "Mail".into(),
        };
        roundtrip(r);
    }

    #[test]
    fn all_request_variants_roundtrip() {
        roundtrip(Request::Insert {
            text: "abc".into(),
            replacement_range: None,
        });
        roundtrip(Request::Insert {
            text: "xyz".into(),
            replacement_range: Some(Range { location: 0, length: 3 }),
        });
        roundtrip(Request::GetContext);
        roundtrip(Request::Ping);
    }

    #[test]
    fn unknown_response_type_fails_with_useful_error() {
        // Defensive: if the Swift side sends a future variant we don't
        // know about, we want a clear deserialize failure, not a panic
        // or a silent miss.
        let r: Result<Response, _> = serde_json::from_str(r#"{"type":"future_thing"}"#);
        assert!(r.is_err(), "expected error, got {r:?}");
    }
}
